datastore/sqlite_datastore.nim

```diff
@@ -339,8 +349,64 @@ method put*(
 
   return await self.put(key, data, timestamp())
 
-# method query*(
-#   self: SQLiteDatastore,
-#   query: ...): Future[?!(?...)] {.async, locks: "unknown".} =
-#
-#   return success ....some
+iterator query*(
+  self: SQLiteDatastore,
+  query: Query): Future[QueryResponse] =
+
+  let
+    queryStmt = QueryStmt.prepare(
+      self.env, queryStmtStr).expect("should not fail")
+
+    s = RawStmtPtr(queryStmt)
+
+  defer:
+    discard sqlite3_reset(s)
+    discard sqlite3_clear_bindings(s)
+    s.dispose
+
+  let
+    v = sqlite3_bind_text(s, 1.cint, query.key.id.cstring, -1.cint,
+      SQLITE_TRANSIENT_GCSAFE)
+
+  if not (v == SQLITE_OK):
+    raise (ref Defect)(msg: $sqlite3_errstr(v))
```

Comment:
As I said before, we should discuss when raise Defect and when return failure, but at this moment let's accept the already implemented approach

Reply:
In this case I wasn't sure of the best approach because we can't `return` from an iterator. We could `yield` a failure and `break` the internal loop, or leave it to the user to `break` in their loop over the iterator when an iteration-step yields a failure. Either way the cleanup in the `defer` block will be triggered.

Since I wasn't sure, for now I chose raise Defect on any error, either directly or via nim-result's `expect`.